src/handlers/proxy.js

```diff
@@ -10,43 +11,38 @@ const {
     Http2OverHttp,
 } = http2.proxies;
 
-/**
- *
- * @param {string|undefined} rawProxyUrl
- * @returns {import('got/dist/source').HandlerFunction} - got handler.
- */
-exports.createProxyHandler = (rawProxyUrl) => {
-    if (rawProxyUrl) {
-        return (options, next) => {
-            options.agent = getAgent(rawProxyUrl, options.http2);
+const proxyHandler = async (options, next) => {
+    const { context: { proxyUrl } } = options;
 
-            return next(options);
-        };
+    if (proxyUrl) {
+        options.agent = await getAgent(proxyUrl, options.https.rejectUnauthorized);
     }
 
-    return (options, next) => next(options);
+    return next(options);
 };
 
 /**
  *
  * @param {string} rawProxyUrl
- * @param {bolean} isHttp2
+ * @param {boolean} rejectUnauthorized
  * @returns {object}
  */
-function getAgent(rawProxyUrl, isHttp2) {
+async function getAgent(rawProxyUrl, rejectUnauthorized) {
     const proxy = {
         proxyOptions: {
             url: new URL(rawProxyUrl),
 
-            rejectUnauthorized: false, // based on the got https.rejectUnauthorized option.
+            rejectUnauthorized, // based on the got https.rejectUnauthorized option.
         },
     };
 
     const proxyUrl = proxy.proxyOptions.url;
-
     let agent;
 
     if (proxyUrl.protocol === 'https:') {
+        const protocol = await httpResolver.resolveHttpVersion(proxyUrl);
+        const isHttp2 = protocol === 'h2';
+
```

Comment:
Do we validate that the `proxyUrl` is only `http` or `https`? This could fail with a weird error because the `else` clause expects HTTP.